In [2]:
import pandas as pd

In [20]:
table_12_2015 = pd.read_excel("data/table_12_agency_hate_crime_reporting_by_state_2015.xls",
                              skiprows=2)
table_12_2015["year"]="2015"
table_12_2015.head(10)

Participating state  Number of\nparticipating\nagencies  \
0                 Total                             14997.0   
1               Alabama                                34.0   
2                Alaska                                33.0   
3               Arizona                               101.0   
4              Arkansas                               279.0   
5            California                               730.0   
6              Colorado                               234.0   
7           Connecticut                                95.0   
8              Delaware                                60.0   
9  District of Columbia                                 2.0   

   Population\ncovered  Agencies\nsubmitting\nincident\nreports  \
0          283884034.0                                   1742.0   
1            1252146.0                                      3.0   
2             734820.0                                      4.0   
3            6622880.0                                     21.0   
4            2754543.0                                      4.0   
5           39137326.0                                    213.0   
6            5445853.0                                     42.0   
7            3399068.0                                     44.0   
8             945934.0                                      7.0   
9             672228.0                                      2.0   

   Total\nnumber of\nincidents\nreported  year  
0                                 5850.0  2015  
1                                   10.0  2015  
2                                    8.0  2015  
3                                  276.0  2015  
4                                    5.0  2015  
5                                  837.0  2015  
6                                  107.0  2015  
7                                   93.0  2015  
8                                   11.0  2015  
9                                   65.0  2015

In [21]:
table_12_2014 = pd.read_excel("data/table_12_agency_hate_crime_reporting_by_state_2014.xls",
                              skiprows=2,parse_cols=4)
table_12_2014["year"]="2014"

table_12_2014.head(10)

Participating state  Number of\nparticipating\nagencies  \
0                 Total                             15494.0   
1               Alabama                                39.0   
2                Alaska                                33.0   
3               Arizona                                99.0   
4              Arkansas                               270.0   
5            California                               743.0   
6              Colorado                               231.0   
7           Connecticut                                95.0   
8              Delaware                                60.0   
9  District of Columbia                                 2.0   

   Population\ncovered  Agencies\nsubmitting\nincident\nreports  \
0          297926030.0                                   1666.0   
1             794907.0                                      4.0   
2             732371.0                                      1.0   
3            6433584.0                                     22.0   
4            2661818.0                                      5.0   
5           38802500.0                                    208.0   
6            5268592.0                                     40.0   
7            3375662.0                                     54.0   
8             935614.0                                      9.0   
9             658893.0                                      2.0   

   Total\nnumber of\nincidents\nreported  year  
0                                 5479.0  2014  
1                                    9.0  2014  
2                                    6.0  2014  
3                                  265.0  2014  
4                                    8.0  2014  
5                                  759.0  2014  
6                                   96.0  2014  
7                                  123.0  2014  
8                                   13.0  2014  
9                                   70.0  2014

In [22]:
table_12_2013 = pd.read_excel("data/table_12_agency_hate_crime_reporting_by_state_2013.xls",
                              skiprows=2,parse_cols=4)
table_12_2013["year"]="2013"

table_12_2013.head(10)

Participating state  Number of\nparticipating\nagencies  \
0                 Total                             15016.0   
1               Alabama                                48.0   
2                Alaska                                33.0   
3               Arizona                                81.0   
4              Arkansas                               265.0   
5            California                               732.0   
6              Colorado                               226.0   
7           Connecticut                                98.0   
8              Delaware                                58.0   
9  District of Columbia                                 2.0   

   Population\ncovered  Agencies\nsubmitting\nincident\nreports  \
0          295016072.0                                   1826.0   
1            1147612.0                                      4.0   
2             730950.0                                      1.0   
3            6358545.0                                     21.0   
4            2809536.0                                     20.0   
5           38324460.0                                    229.0   
6            5163276.0                                     48.0   
7            3566367.0                                     48.0   
8             925749.0                                     11.0   
9             646449.0                                      2.0   

   Total\nnumber of\nincidents\nreported  year  
0                                 5928.0  2013  
1                                    6.0  2013  
2                                    8.0  2013  
3                                  155.0  2013  
4                                   27.0  2013  
5                                  843.0  2013  
6                                  128.0  2013  
7                                  145.0  2013  
8                                   12.0  2013  
9                                   72.0  2013

In [38]:
table_12_2012 = pd.read_excel("data/table_12_agency_hate_crime_reporting_by_state_2012.xls",
                              skiprows=2,parse_cols=4)
table_12_2012["year"]="2012"
table_12_2012.head(10)

Index([                    u'Participating state',
            u'Number of\nparticipating\nagencies',
                           u'Population\ncovered',
       u'Agencies\nsubmitting\nincident\nreports',
         u'Total\nnumber of\nincidents\nreported',
                                          u'year'],
      dtype='object')


Participating state  Number of\nparticipating\nagencies  \
0                 Total                             13022.0   
1               Alabama                                 8.0   
2                Alaska                                 5.0   
3               Arizona                                81.0   
4              Arkansas                               265.0   
5            California                               731.0   
6              Colorado                               227.0   
7           Connecticut                               101.0   
8              Delaware                                57.0   
9  District of Columbia                                 2.0   

   Population\ncovered  Agencies\nsubmitting\nincident\nreports  \
0          248809710.0                                   1730.0   
1             141235.0                                      1.0   
2             315266.0                                      1.0   
3            5974180.0                                     22.0   
4            2748544.0                                     17.0   
5           37933751.0                                    247.0   
6            5109141.0                                     50.0   
7            3574937.0                                     48.0   
8             917092.0                                      5.0   
9             632323.0                                      2.0   

   Total\nnumber of\nincidents\nreported  year  
0                                 5796.0  2012  
1                                    6.0  2012  
2                                    6.0  2012  
3                                  176.0  2012  
4                                   30.0  2012  
5                                  910.0  2012  
6                                  189.0  2012  
7                                  149.0  2012  
8                                   10.0  2012  
9                                   72.0  2012

In [39]:
table_12_2011 = pd.read_excel("data/table-12-2011.xls",
                              skiprows=2)
table_12_2011["year"]="2011"
print table_12_2012.columns

table_12_2011.head(10)

Index([                    u'Participating state',
            u'Number of\nparticipating\nagencies',
                           u'Population\ncovered',
       u'Agencies\nsubmitting\nincident\nreports',
         u'Total\nnumber of\nincidents\nreported',
                                          u'year'],
      dtype='object')


Participating state  Number of\nparticipating\nagencies  \
0                 Total                             14575.0   
1               Alabama                                98.0   
2                Alaska                                 2.0   
3               Arizona                                84.0   
4              Arkansas                               259.0   
5            California                               734.0   
6              Colorado                               218.0   
7           Connecticut                               102.0   
8              Delaware                                56.0   
9  District of Columbia                                 2.0   

   Population\ncovered  Agencies\nsubmitting\nincident\nreports  \
0          286010550.0                                   1944.0   
1            2285487.0                                     50.0   
2             304924.0                                      1.0   
3            6225041.0                                     26.0   
4            2804225.0                                     10.0   
5           37691912.0                                    257.0   
6            4992496.0                                     57.0   
7            3580709.0                                     42.0   
8             907135.0                                      8.0   
9             617996.0                                      2.0   

   Total\nnumber of\nincidents\nreported  year  
0                                 6222.0  2011  
1                                   83.0  2011  
2                                    8.0  2011  
3                                  192.0  2011  
4                                   11.0  2011  
5                                 1040.0  2011  
6                                  186.0  2011  
7                                  140.0  2011  
8                                   15.0  2011  
9                                   83.0  2011

In [41]:
table_12_2010 = pd.read_excel("data/table-12-agency-hate-crime-reporting-by-state-2010.xls",
                              skiprows=2)
table_12_2010["year"]="2010"
print table_12_2010.columns
table_12_2010.columns = ['Participating state',
                         'Number of\nparticipating\nagencies',
                         'Population\ncovered',
                         'Agencies\nsubmitting\nincident\nreports',
                         'Total\nnumber of\nincidents\nreported',
                         "year"]
table_12_2010.head(10)

Index([                     u'Participating state',
             u'Number of\nparticipating\nagencies',
                            u'Population\ncovered',
       u'Agencies\nsubmitting\n incident\nreports',
           u'Total number\nof incidents\nreported',
                                           u'year'],
      dtype='object')


Participating state  Number of\nparticipating\nagencies  \
0                 Total                               14977   
1               Alabama                                 331   
2                Alaska                                   1   
3               Arizona                                  89   
4              Arkansas                                 254   
5            California                                 733   
6              Colorado                                 217   
7           Connecticut                                 102   
8              Delaware                                  56   
9  District of Columbia                                   2   

   Population\ncovered  Agencies\nsubmitting\nincident\nreports  \
0            285001266                                     1949   
1              3762738                                       12   
2               290334                                        1   
3              6328446                                       22   
4              2752987                                       30   
5             37200288                                      249   
6              4908638                                       44   
7              3574097                                       43   
8               897934                                       10   
9               601723                                        1   

   Total\nnumber of\nincidents\nreported  year  
0                                   6628  2010  
1                                     19  2010  
2                                      7  2010  
3                                    236  2010  
4                                     63  2010  
5                                   1092  2010  
6                                    154  2010  
7                                    147  2010  
8                                     20  2010  
9                                     57  2010

In [42]:
# Check that all column headings are already the same
print table_12_2010.columns
print table_12_2010.columns == table_12_2011.columns
print table_12_2011.columns == table_12_2012.columns
print table_12_2012.columns == table_12_2013.columns
print table_12_2013.columns == table_12_2014.columns
print table_12_2014.columns == table_12_2015.columns

Index([u'Participating state', u'Number of\nparticipating\nagencies',
       u'Population\ncovered', u'Agencies\nsubmitting\nincident\nreports',
       u'Total\nnumber of\nincidents\nreported', u'year'],
      dtype='object')
[ True  True  True  True  True  True]
[ True  True  True  True  True  True]
[ True  True  True  True  True  True]
[ True  True  True  True  True  True]
[ True  True  True  True  True  True]


In [43]:
print table_12_2010[table_12_2010["Participating state"] == "Connecticut"]["Population\ncovered"]
print table_12_2011[table_12_2011["Participating state"] == "Connecticut"]["Population\ncovered"]
print table_12_2012[table_12_2012["Participating state"] == "Connecticut"]["Population\ncovered"]
print table_12_2013[table_12_2013["Participating state"] == "Connecticut"]["Population\ncovered"]
print table_12_2014[table_12_2014["Participating state"] == "Connecticut"]["Population\ncovered"]
print table_12_2015[table_12_2015["Participating state"] == "Connecticut"]["Population\ncovered"]


7    3574097
Name: Population\ncovered, dtype: int64
7    3580709.0
Name: Population\ncovered, dtype: float64
7    3574937.0
Name: Population\ncovered, dtype: float64
7    3566367.0
Name: Population\ncovered, dtype: float64
7    3375662.0
Name: Population\ncovered, dtype: float64
7    3399068.0
Name: Population\ncovered, dtype: float64


In [44]:
#### IMPORTANT
# Because the number of agencies reporting in a given year vary, but the "population covered"
# column in this table is the whole state population, i instead want to use the total number 
# of populations per agency based on Table 13. However, that's skewed as well, since it does
# not include towns that participated but reported no hate crimes, which was about half of 
# all agencies. Both methodologies are imperfect.

# Calculations from my other sheet: 
#2014 2210928.0
#2011 1785489.0
#2010 1638134.0
#2013 1970207.0
#2012 1974862.0

effective_pops = {'2010': 1638134.0,
 '2011': 1785489.0,
 '2012': 1974862.0,
 '2013': 1970207.0,
 '2014': 2210928.0,
 '2015_unadjusted': 2346897.0,
 '2015': 1822339.0}


In [45]:
def effective_per_capita(df):
    tmp_df = df.copy();
    year = tmp_df["year"].max()
    tmp_df["adj_per_cap"] = tmp_df["Total\nnumber of\nincidents\nreported"] * 1000 * 1000 / effective_pops[year]

    tmp_df["adj_per_cap_rank"] = tmp_df["adj_per_cap"].rank(ascending=False)
    return tmp_df.sort_values(by="adj_per_cap",ascending=False)

def per_capita (df):
    tmp_df = df.copy();
    tmpdf = effective_per_capita(tmp_df)
    tmp_df["per_cap"] = tmp_df["Total\nnumber of\nincidents\nreported"] * 1000 * 1000 / tmp_df["Population\ncovered"]

    tmp_df["per_cap_rank"] = tmp_df["per_cap"].rank(ascending=False)
    
    return tmp_df.sort_values(by="per_cap",ascending=False)

def state_only(df, state):
    return df[df["Participating state"]==state]


def ct_only(df):
    return state_only(df, "Connecticut")

def agencies_participating_only(df):
    return df["Number of\nparticipating\nagencies"]

def agencies_reporting_only(df):
    return df["Agencies\nsubmitting\nincident\nreports"]

def percap_only(df):
    return df["per_cap"]

def rank_only(df):
    return df["per_cap_rank"]

def state_rank(df,state):
    return rank_only(state_only(df,state))

def state_percap(df,state):
    return percap_only(state_only(df,state))

def ct_rank(df):
    return int(rank_only(state_only(per_capita(df),"Connecticut")))

def ct_percap(df):
    return int(percap_only(state_only(per_capita(df),"Connecticut")))

def ct_participating(df):
    return int(agencies_participating_only(state_only(per_capita(df),"Connecticut")))

def ct_reporting(df):
    return int(agencies_reporting_only(state_only(per_capita(df),"Connecticut")))

In [46]:
print int(ct_only(table_12_2010)['Total\nnumber of\nincidents\nreported'])
print int(ct_only(table_12_2011)['Total\nnumber of\nincidents\nreported'])
print int(ct_only(table_12_2012)['Total\nnumber of\nincidents\nreported'])
print int(ct_only(table_12_2013)['Total\nnumber of\nincidents\nreported'])
print int(ct_only(table_12_2014)['Total\nnumber of\nincidents\nreported'])
print int(ct_only(table_12_2015)['Total\nnumber of\nincidents\nreported'])

147
140
149
145
123
93


In [47]:
print ct_rank(table_12_2010)
print ct_rank(table_12_2011)
print ct_rank(table_12_2012)
print ct_rank(table_12_2013)
print ct_rank(table_12_2014)
print ct_rank(table_12_2015)

8
8
7
6
9
15


In [48]:
print ct_percap(table_12_2010)
print ct_percap(table_12_2011)
print ct_percap(table_12_2012)
print ct_percap(table_12_2013)
print ct_percap(table_12_2014)
print ct_percap(table_12_2015)

41
39
41
40
36
27


In [49]:
print ct_participating(table_12_2010)
print ct_participating(table_12_2011)
print ct_participating(table_12_2012)
print ct_participating(table_12_2013)
print ct_participating(table_12_2014)
print ct_participating(table_12_2015)

102
102
101
98
95
95


In [50]:
print ct_reporting(table_12_2010)
print ct_reporting(table_12_2011)
print ct_reporting(table_12_2012)
print ct_reporting(table_12_2013)
print ct_reporting(table_12_2014)
print ct_reporting(table_12_2015)

43
42
48
48
54
44


### CT conclusions from table 12

Based on the above analysis: 

The number and rate of reported hate crime incidents in CT were at a low in 2015 since data became available in 2010 and the state's rank improved over that time. However, the number of police agencies participating dropped from 102 agencies to 95 in Connecticut. The number of agencies reporting incidents did not decline over that period. The number of agencies reporting hate crime incidents ranged from 42 to 54 agencies in 2014. In 2015, it was 44.


### But...

As described above, it turns out the per_capita information from the feds is based on the entire state population, even though about 100 departments participate in this reporting program and only about 40-50 towns report at least hate crime. And because the number of agencies reporting changes from year to year, we don't have a constant baseline.

In [51]:
table_12 = {
    "2010":table_12_2010,
    "2011":table_12_2011,
    "2012":table_12_2012,
    "2013":table_12_2013,
    "2014":table_12_2014,
    "2015":table_12_2015
}

In [52]:
def totals_obj():
    ret = {}
    for year in [2010,2011,2012,2013,2014,2015]:
        ret[str(year)] = int(ct_only(table_12[str(year)])["Total\nnumber of\nincidents\nreported"])
    return ret
totals_obj()

{'2010': 147, '2011': 140, '2012': 149, '2013': 145, '2014': 123, '2015': 93}